### ISSUE TITLE
.NET 8 Benchmarks are likely testing "cold"

### ISSUE THREAD

**Issue by tannergooding** [tbg10101/dotnet-burst-comparison#1]
The readme notes:
> Each benchmark is run twice to let each runtime warm up. These results are from the second run. Smaller values are better.

However, modern .NET (.NET Core 2.1+) uses "Tiered Compilation" and that defaults to needing approx. 30 invocations and a delay of 1ms where no other Tier-0 code is jitted before methods can start being promoted to Tier-1

The very latest versions (.NET 7+) do have some support for OSR (On Stack Replacement) in which a hot loop may be replaced even when the rest of the method body hasn't been optimized and I expect that's what's happening for these benchmarks (as many do have loops that execute far more than 30 invocations). However, that is still going to miss some opts and pessimize the overall codegen.

In general, I'd recommend either:
* disabling `TieredCompilation` (https://learn.microsoft.com/en-us/dotnet/core/runtime-config/compilation)
* ensuring that a minimum of 30-50 invocations occur and a sufficient pause time (1-5ms) exists before measuring to allow rejit to kick in
* just use `Benchmark.NET` which will handle all of this for you, ensuring enough invocations occur and the relevant statistical analysis is done to find and account for noise, outlier results, etc.

If you go with the first approach (disabling TC), it's worth noting that this also opts you out of things like DynamicPGO and optimizations that can only occur via a TieredJIT where you know some code has executed at least once.

**Comment by tannergooding** [comment]
Separately, it's worth noting that there are some pessimizations that exist here which could be accounted for...

A simple example is in `SieveOfEratosthenes` where instead of:
```csharp
byte* flags = stackalloc byte[size];

// ....

for (b = 0; b < size; b++) {
    flags[b] = 1; // True
}
```

the code should probably just do:
```csharp
Span<byte> flags = stackalloc byte[size];

// ...

flags.Fill(1);
```

Doing cleanup like this will improve code quality and legibility/maintainability. It's really the same kind of scenarios where you'd use `memset` or similar built-in functions in C/C++.

**Comment by tbg10101** [comment]
I'll see about using Benchmark.NET today. (for both projects, for consistency)

For the code changes, could you create a separate issue for that? (if not, that's fine, I can) Sadly I don't have the background for why things were done the way they were in the [predecessor project](https://github.com/nxrighthere/BurstBenchmarks) so I mostly left things alone. But I agree that the changes are warranted. I'd like each version to be the best it can be for the framework being tested.

**Comment by tbg10101** [comment]
The migration for the vanilla .NET project was painless but I didn't know it [doesn't work for Unity](https://github.com/dotnet/BenchmarkDotNet/issues/1860). So I parked that work on a branch called [benchmark.net](https://github.com/tbg10101/dotnet-burst-comparison/tree/benchmark.net).

I'll implement my own cheap version as outlined here: https://benchmarkdotnet.org/articles/guides/how-it-works.html

**Comment by tbg10101** [comment]
I've run into a strange issue with the Burst Sieve Of Eratosthenes after changing it to `stackalloc` a `byte[]` and use `Fill()`. It seems to hang the runtime. Unfortunately this is hard to debug because it works fine if I drop a breakpoint into the job.

For example, this works:

```csharp
const int size = 1024;

Span<byte> flags = stackalloc byte[size];
flags.Fill(1); // True

return (uint)flags.Length;
```

But this hangs forever:

```csharp
const int size = 1024;

Span<byte> flags = stackalloc byte[size];
flags.Fill(1); // True

return flags[index];
```

It is either the fill or the array lookup causing the issue. (the fill gets optimized out of the first one)

This post from 2021 says Burst should have full Span support but perhaps it is an issue with the backing byte array: https://forum.unity.com/threads/burstcompile-span-t-support.603310/#post-7544536

I'll switch it to use a `NativeArray<byte>` which can also be `Fill()`-ed.

**Comment by tbg10101** [comment]
I just pushed a big change to address your comments. Let me know if there is anything else I should do.

I have re-run the benchmark but some things don't match my expectations. For example IL2CPP without Burst is worse than Mono (also without Burst) in the Fibonacci benchmark.

I need to investigate irregularities but I'm going on vacation tomorrow so I figured I'd push what I had.

**Comment by tbg10101** [comment]
I think I figured out the issue - I found that a background task would start if I left my computer idle while the benchmarks ran. This wasn't an issue in the previous iteration since they only ran once and were quite fast. In the new version the tests are run multiple times so the overall testing time has expanded, leaving time for tasks that detect idleness to kick in.

I have re-run the benchmarks without letting my computer become idle and will publish the results later today. This issue will be closed after that.
